### ISSUE TITLE
Wrong inference calculation

### ISSUE THREAD

**Issue by Chaoste** [bayespy/bayespy#94]
Hi,

I'm not sure if there is a problem with solving the inference task or I'm using the module the wrong way. I have to implement the [sprinkler example by Kevin Murphy](http://people.cs.ubc.ca/~murphyk/Bayes/bnintro.html).
After implementing and executing inference I get wrong results. You can find my code [here](https://github.com/Chaoste/course-machine-intelligence-1/blob/master/Ex10/MI%20-%20H10.ipynb).
The results of my code:
```
P(s=1 | w=1) = 0.148459431093
P(r=1 | w=1) = 1.0
```
The right values are according to the calculations by Kevin Murphy:
```
P(s=1 | w=1) = 0.430
P(r=1 | w=1) = 0.708
```

Greetings,
Thomas

**Comment by jluttine** [comment]
Yep, BayesPy uses variational Bayesian (VB) approximation for the graph. It doesn't make much sense for such a small graph, but the exact inference algorithms for discrete graphs are under development currently. So, basically you'll get incorrect (i.e., approximate) results in this case.

The used VB approximation assumes that the variables are independent in the posterior. However, this is not the case in general and in particular not the case in this example. Because you observe "grass is wet" and you have `P(grass is wet | no sprinkler, no rain) = 0`, either "sprinkler" or "rain" has to be "true". But because VB doesn't model the coupling, it chooses "rain" to be "true" because it's a more probable explanation. A rule of thumb: if you have probabilities near 1.0 in the conditional probability tables, there's a risk that the VB approximation can be really bad.

We are currently working on exact inference for discrete graphs and I hope it will become available in the near future: https://github.com/bayespy/bayespy/issues/37

There is a "brute-force" way of solving this problem. You could model the latent variables jointly with one variable with 2^3=8 possible values/states. Then you'll not only get true marginals but you'll get the true full joint posterior. Here's how to do it.


```python
from bayespy.nodes import Categorical, Mixture, MultiMixture
from bayespy.inference import VB

cloudy = [0.5, 0.5] # p(cloudy) = 0.5
rain = [
    [0.8, 0.2], # p(rain | not cloudy) = 0.2
    [0.2, 0.8], # p(rain | cloudy) = 0.8
]
sprinkler = [
    [0.5, 0.5], # p(sprinkler on | not cloudy) = 0.5
    [0.9, 0.1], # p(sprinkler on | cloudy) = 0.1
]

joint_prob = np.einsum('c,cr,cs->crs', cloudy, rain, sprinkler)

joint = Categorical(joint_prob.ravel())

likelihood_prob = np.array(
    [
        # Not cloudy
        [
            # Not rain
            [
                # Not sprinkler
                [1.0, 0.0],
                # Sprinkler
                [0.1, 0.9]
            ],
            # Rain
            [
                # Not sprinkler
                [0.1, 0.9],
                # Sprinkler
                [0.01, 0.99]
            ]
        ],        
        # Cloudy
        [
            # Not rain
            [
                # Not sprinkler
                [1.0, 0.0],
                # Sprinkler
                [0.1, 0.9]
            ],
            # Rain
            [
                # Not sprinkler
                [0.1, 0.9],
                # Sprinkler
                [0.01, 0.99]
            ]
        ],        
    ]
)

wetGrass = Mixture(
    joint,
    Categorical,
    likelihood_prob.reshape((-1, 2))
)

wetGrass.observe(1)

Q = VB(wetGrass, joint)
Q.update(repeat=100, verbose=True)

# Reshape the posterior probability table so that each variable has its own axis.
joint_posterior = joint.get_moments()[0].reshape((2,2,2))

# Show marginals
print('P(c=1|w=1) =', np.einsum('crs->c', joint_posterior)[1])
print('P(r=1|w=1) =', np.einsum('crs->r', joint_posterior)[1])
print('P(s=1|w=1) =', np.einsum('crs->s', joint_posterior)[1])
```

    Iteration 1: loglike=-4.352544e-01 (0.001 seconds)
    Iteration 2: loglike=-4.352544e-01 (0.001 seconds)
    Converged at iteration 2.
    P(c=1|w=1) = 0.575799721836
    P(r=1|w=1) = 0.70792767733
    P(s=1|w=1) = 0.429763560501


    /home/jluttine/Workspace/bayespy/bayespy/inference/vmp/nodes/dirichlet.py:91: RuntimeWarning: divide by zero encountered in log
      logp = np.log(p)


Note that now we can print the joint probability distribution of "rain" and "sprinkler". And notice that the probability p(no rain, no sprinkler|wet grass) is zero.


```python
print("Joint probability p(rain,sprinkler|wet grass):")
print(np.einsum('crs->rs', joint_posterior))
```

    Joint probability p(rain,sprinkler|wet grass):
    [[ 0.          0.29207232]
     [ 0.57023644  0.13769124]]


Did this answer your question? If there was something unclear please don't hesitate to ask for more detailed explanation.

**Comment by jluttine** [comment]
Of course, you don't need to use BayesPy at all for such a simple problem. You can use, for instance, numpy.einsum directly to compute the required posterior distributions conveniently. Just write all conditional probability tables as NumPy arrays where each axis corresponds to one of the variables. Then you multiply them all together to get the full joint probability table, slice the "likelihood axis" to condition on the observation, sum over the axis you want to marginalize over and normalize.

```
# Conditional probability tables
cloudy = [0.5, 0.5] # p(cloudy) = 0.5
rain = [
    [0.8, 0.2], # p(rain | not cloudy) = 0.2
    [0.2, 0.8], # p(rain | cloudy) = 0.8
]
sprinkler = [
    [0.5, 0.5], # p(sprinkler on | not cloudy) = 0.5
    [0.9, 0.1], # p(sprinkler on | cloudy) = 0.1
]
grass = [
    # Not rain
    [
        # Not sprinkler
        [1.0, 0.0],
        # Sprinkler
        [0.1, 0.9]
    ],
    # Rain
    [
        # Not sprinkler
        [0.1, 0.9],
        # Sprinkler
        [0.01, 0.99]
    ]
]

# Joint probability
joint = np.einsum('c,cr,cs,rsg->crsg', cloudy, rain, sprinkler, grass)

# Observe wet grass (pick the correct slice from the grass axis)
joint_posterior = joint[...,1]

def normalize(p):
    return p / np.sum(p)

# Compute marginal
print("Cloudy:", normalize(np.einsum('crs->c', joint_posterior))[1])
print("Sprinkler:", normalize(np.einsum('crs->r', joint_posterior))[1])
print("Grass:", normalize(np.einsum('crs->s', joint_posterior))[1])

# Compute some joint posteriors
print("Rain and sprinkler:")
print(normalize(np.einsum('crs->rs', joint_posterior)))
```

This method (and the above method with BayesPy I suggested) don't scale to large graphs because these methods compute the full joint distributions and these arrays grow exponentially with the size of the graph.
